File: src/main/java/mil/nga/tiff/field/type/enumeration/DifferencingPredictor.java
package mil.nga.tiff.field.type.enumeration;

import mil.nga.tiff.util.TiffException;

import java.util.Arrays;

public enum DifferencingPredictor {
    NO(1),
    HORIZONTAL(2),
    FLOATINGPOINT(3);

    private final int id;

    DifferencingPredictor(int id) {
        this.id = id;
    }

    public static DifferencingPredictor findById(int id) {
        return Arrays.stream(values())
            .filter(o -> o.id == id)
            .findAny()
            .orElseThrow(() -> new TiffException("Invalid ID: " + id));
    }

    public int getId() {
        return id;
    }
}

File: src/main/java/mil/nga/tiff/field/type/enumeration/DifferencingPredictor.kt
package mil.nga.tiff.field.type.enumeration

import mil.nga.tiff.internal.predictor.FloatingPointPredictor
import mil.nga.tiff.internal.predictor.HorizontalPredictor
import mil.nga.tiff.internal.predictor.NullPredictor
import mil.nga.tiff.internal.predictor.Predictor

enum class DifferencingPredictor(val id: Int, val implementation: Predictor) {
    NO(1, NullPredictor()),
    HORIZONTAL(2, HorizontalPredictor()),
    FLOATINGPOINT(3, FloatingPointPredictor());

    companion object {
        private val DEFAULT = NO

        @JvmStatic
        fun findById(id: Int?): DifferencingPredictor {
            if (id == null) {
                return DEFAULT
            }

            return entries.first { it.id == id}
        }
    }
}
